arch/x86/entry.S

```diff
@@ -153,6 +154,16 @@ ENTRY(enter_usermode)
     IRET
 END_FUNC(enter_usermode)
 
+ENTRY(terminate_user_task)
+    SWITCH_STACK
+    POPF
+
+    movl $-EFAULT, cpu_regs_ax(%_ASM_SP)
+    RESTORE_ALL_REGS
+
+    ret
```

Comment:
Yes, this is intentional. We are already in the kernel, but because the exception happened in usermode we kept the user stack, displayed the exception in `do_exception()` and jumped to this function. Now the function switches the stack to kernel one, and undoes what `enter_usermode()` did (pops flags, saves return code into `AX`, which is -EFAULT for exceptions in usermode and restores previously saved kernel regs). The `RET` returns to the caller of `enter_usermode()`.
Which is the user task creation function.

Reply:
Ahh, thanks for the explanation!